os/arch/arm/src/amebasmart/amebasmart_mipi.c

```diff
@@ -0,0 +1,355 @@
+	u32 totalx = MIPI_DSI_HSA + MIPI_DSI_HBP + MIPI_DSI_HFP + data->XPixels;
+	vo_freq = totaly * totalx * MIPI_FRAME_RATE / Mhz + 4;
+	dbg("vo_freq: %d\n", vo_freq);
+	assert_param(vo_freq < 67);
+}
+
+static void amebasmart_set_clock(void)
+{
+	u32 PLLDiv = PLL_GET_NPLL_DIVN_SDM(PLL_BASE->PLL_NPPLL_CTRL1) + 2;
+	u32 PLL_CLK = XTAL_ClkGet() * PLLDiv;
+	u32 mipi_ckd = PLL_CLK / vo_freq - 1;
+	HAL_WRITE32(SYSTEM_CTRL_BASE_LP, REG_LSYS_CKD_GRP0, (HAL_READ32(SYSTEM_CTRL_BASE_LP, REG_LSYS_CKD_GRP0) & ~LSYS_MASK_CKD_MIPI) | LSYS_CKD_MIPI(mipi_ckd));
+	HAL_WRITE32(SYSTEM_CTRL_BASE_LP, REG_LSYS_CKD_GRP0, (HAL_READ32(SYSTEM_CTRL_BASE_LP, REG_LSYS_CKD_GRP0) & ~LSYS_MASK_CKD_HPERI) | LSYS_CKD_HPERI(3));
+
+	RCC_PeriphClockCmd(APBPeriph_NULL, APBPeriph_HPERI_CLOCK, ENABLE);
+	RCC_PeriphClockCmd(APBPeriph_LCDC, APBPeriph_LCDCMIPI_CLOCK, ENABLE);
+
+}
+
+static void ameabsmart_register_interrupt(void)
+{
+	InterruptDis(mipi_irq_info.num);
+	InterruptUnRegister(mipi_irq_info.num);
+	InterruptRegister((IRQ_FUN) amebasmart_mipidsi_isr, mipi_irq_info.num, (u32) mipi_irq_info.data, mipi_irq_info.priority);
+	InterruptEn(mipi_irq_info.num, mipi_irq_info.priority);
+}
+
+static bool ameabsmart_is_transfer_completed(void)
+{
+	if (send_cmd_done == 1) {
+		return true;
+	}
+	return false;
+}
+
+void mipidsi_mode_switch(bool do_enable){
+	if(do_enable){
+		MIPI_DSI_Mode_Switch(g_dsi_host.MIPIx, ENABLE);
+	}
+	else{
+		MIPI_DSI_Mode_Switch(g_dsi_host.MIPIx, DISABLE);
+	}
+}
+
+void mipidsi_acpu_reg_clear(void){
+	MIPI_DSI_INTS_ACPU_Clr(g_dsi_host.MIPIx, MIPI_DSI_INTS_ACPU_Get(g_dsi_host.MIPIx));
+}
+
+/************************************************************************************
+ * Private Functions
+ ************************************************************************************/
+static int amebasmart_mipi_attach(FAR struct mipi_dsi_host *dsi_host, FAR struct mipi_dsi_device *dsi_device)
+{
+	return 0;
+}
+
+static int amebasmart_mipi_detach(FAR struct mipi_dsi_host *dsi_host, FAR struct mipi_dsi_device *dsi_device)
+{
+	return 0;
+}
+
+static int amebasmart_mipi_transfer(FAR struct mipi_dsi_host *dsi_host, FAR const struct mipi_dsi_msg *msg)
+{
+	FAR struct amebasmart_mipi_dsi_host_s *priv = (FAR struct amebasmart_mipi_dsi_host_s *)dsi_host;
+	struct mipi_dsi_packet packet;
+
+	if(msg->type == MIPI_DSI_END_OF_TRANSMISSION){
+		MIPI_DSI_INT_Config(g_dsi_host.MIPIx, DISABLE, DISABLE, FALSE);
+		return OK;
+	}
+	
+	int ret = mipi_dsi_create_packet(&packet, msg);
+	if (ret != OK) {
+		return ret;
+	}
+	send_cmd_done = 0;
+	if (mipi_dsi_packet_format_is_short(msg->type)) {
+		if (packet.header[1] == 0) {
+			amebasmart_mipidsi_send_cmd(priv->MIPIx, packet.header[0], 0, NULL);
+		} else {
+			amebasmart_mipidsi_send_cmd(priv->MIPIx, packet.header[0], 1, packet.header + 1);
+		}
+	} else {
+		amebasmart_mipidsi_send_cmd(priv->MIPIx, packet.header[0], packet.payload_length, packet.payload);
+	}
+
+	while (!ameabsmart_is_transfer_completed()) {
```

Comment:
This seems having much delay (ie. 1ms), are we supposed to wait for such long time? Just polling is not enough?
The intention of adding delay here seems that we are sending a table of command list during initialization stage. But if we are just sending command individually, perhaps the delay is unnecessary.

Reply:
I have removed the delay and changed the while loop to this:

	while(true){
		if(send_cmd_done == 1){
			break;
		}
		count++;
		if(count >=20){
			dbg("Error: amebasmart mipi send cmd timeout\n");
		}
	}

Reply:
I have 2 questions:
1) How do you get the value of 20?
2) What if send_cmd_done never gets updated? Maybe we should add some error handling instead of just printing timeout logs (ie. return false?)

Reply:
Hello, I have changed this code. It seems that it's not possible to exit from while loop without adding sleep. It gets stuck in the while loop. Not sure about the reason. I have added the 1ms delay. It's the lowest I could go when I ran the code in hardware.
The reason it was working before is that I have added some debug logs before the while loop. It seems that debug log results in some delay. When I remove the debug log just before the while loop, then code gets stuck in the while loop.

The new code is working properly , ... I verified many times.